File: android/ui-framework-sample/ui-framework/src/main/java/kh/mcnc/lib/components/BizCalendar.kt
package kh.mcnc.lib.components

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.ExperimentalAnimationApi
import androidx.compose.animation.SizeTransform
import androidx.compose.animation.core.tween
import androidx.compose.animation.slideInHorizontally
import androidx.compose.animation.slideOutHorizontally
import androidx.compose.animation.with
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ChevronLeft
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.Today
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kh.mcnc.lib.utils.CalendarConfig
import kh.mcnc.lib.utils.CalendarDay
import kh.mcnc.lib.utils.CalendarState
import kh.mcnc.lib.utils.SelectionMode
import java.util.Calendar
import java.util.Date

@OptIn(ExperimentalAnimationApi::class)
@Composable
fun BizCalendar(
    config: CalendarConfig = CalendarConfig(),
    initialDate: Date = Date(),
    showActionButtons: Boolean = false,
    onDateSelected: ((Date) -> Unit)? = null,
    onRangeSelected: ((start: Date?, end: Date?) -> Unit)? = null,
    onMonthChanged: ((year: Int, month: Int) -> Unit)? = null,
    onCancel: (() -> Unit)? = null,
    onConfirm: ((start: Date?, end: Date?) -> Unit)? = null
) {
    val primaryColor = Color(0xFF1A2A5E)
    val dangerColor = Color(0xFFFF4D4F)
    val rangeBgColor = dangerColor.copy(alpha = 0.1f)
    
    var currentMonth by remember(initialDate) { mutableStateOf(Calendar.getInstance().apply { time = initialDate }.get(Calendar.MONTH) + 1) }
    var currentYear by remember(initialDate) { mutableStateOf(Calendar.getInstance().apply { time = initialDate }.get(Calendar.YEAR)) }
    
    var selectedDate by remember(initialDate) { mutableStateOf<Date?>(initialDate) }
    var rangeStartDate by remember { mutableStateOf<Date?>(null) }
    var rangeEndDate by remember { mutableStateOf<Date?>(null) }
    
    val calendarState = remember(config) { CalendarState(config) }
    var days by remember { mutableStateOf(emptyList<CalendarDay>()) }
    var isLoading by remember { mutableStateOf(false) }
    
    var transitionDirection by remember { mutableStateOf(1) } // 1 for next, -1 for prev
    
    val monthNames = listOf(
        "Jan", "Feb", "Mar", "Apr", "May", "Jun", 
        "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
    )
    
    val daysOfWeek = if (config.firstDayOfWeek == 1) {
        listOf("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun")
    } else {
        listOf("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")
    }

    LaunchedEffect(currentMonth, currentYear, calendarState) {
        isLoading = true
        kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.Default) {
            val newDays = calendarState.generateDaysForMonth(currentYear, currentMonth)
            days = newDays
        }
        isLoading = false
        onMonthChanged?.invoke(currentYear, currentMonth)
    }

    fun isSameDay(d1: Date?, d2: Date?): Boolean {
        if (d1 == null || d2 == null) return false
        val cal1 = Calendar.getInstance().apply { time = d1 }
        val cal2 = Calendar.getInstance().apply { time = d2 }
        return cal1.get(Calendar.YEAR) == cal2.get(Calendar.YEAR) &&
               cal1.get(Calendar.MONTH) == cal2.get(Calendar.MONTH) &&
               cal1.get(Calendar.DAY_OF_MONTH) == cal2.get(Calendar.DAY_OF_MONTH)
    }

    fun nextMonth() {
        transitionDirection = 1
        if (currentMonth == 12) {
            currentMonth = 1
            currentYear += 1
        } else {
            currentMonth += 1
        }
    }

    fun prevMonth() {
        transitionDirection = -1
        if (currentMonth == 1) {
            currentMonth = 12
            currentYear -= 1
        } else {
            currentMonth -= 1
        }
    }

    fun goToToday() {
        transitionDirection = 1
        val today = Calendar.getInstance()
        currentMonth = today.get(Calendar.MONTH) + 1
        currentYear = today.get(Calendar.YEAR)
    }

    fun onDayClick(day: CalendarDay) {
        if (!day.isEnabled || !day.isCurrentMonth) return
        
        when (config.selectionMode) {
            SelectionMode.Single -> {
                selectedDate = day.date
                onDateSelected?.invoke(day.date)
            }
            SelectionMode.Week -> {
                val index = days.indexOfFirst { isSameDay(it.date, day.date) }
                if (index != -1) {
                    val rowStart = (index / 7) * 7
                    val weekDays = days.subList(rowStart, Math.min(rowStart + 7, days.size)).filter { it.isEnabled && !it.isHidden }
                    if (weekDays.isNotEmpty()) {
                        rangeStartDate = weekDays.first().date
                        rangeEndDate = weekDays.last().date
                        onRangeSelected?.invoke(rangeStartDate, rangeEndDate)
                    }
                }
            }
            SelectionMode.Range -> {
                if (rangeStartDate == null || (rangeStartDate != null && rangeEndDate != null)) {
                    rangeStartDate = day.date
                    rangeEndDate = null
                } else if (rangeStartDate != null && rangeEndDate == null) {
                    if (day.date.time < rangeStartDate!!.time) {
                        rangeStartDate = day.date
                    } else {
                        rangeEndDate = day.date
                        onRangeSelected?.invoke(rangeStartDate, rangeEndDate)
                    }
                }
            }
        }
    }

    Column(modifier = Modifier
        .fillMaxWidth()
        .background(Color.White)
        .padding(16.dp)) {
        
        // Header
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 16.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = "${monthNames[currentMonth - 1]} $currentYear",
                fontSize = 18.sp,
                fontWeight = FontWeight.SemiBold,
                color = primaryColor,
                modifier = Modifier.clickable { /* TODO: Show Month Picker */ }
            )
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                IconButton(onClick = { prevMonth() }) {
                    Icon(Icons.Default.ChevronLeft, contentDescription = "Previous Month", tint = primaryColor)
                }
                IconButton(onClick = { goToToday() }) {
                    Icon(Icons.Default.Today, contentDescription = "Today", tint = primaryColor)
                }
                IconButton(onClick = { nextMonth() }) {
                    Icon(Icons.Default.ChevronRight, contentDescription = "Next Month", tint = primaryColor)
                }
            }
        }

        // Weekdays
        Row(modifier = Modifier.fillMaxWidth().padding(bottom = 8.dp)) {
            daysOfWeek.forEachIndexed { index, dayName ->
                val isSunday = if (config.firstDayOfWeek == 1) index == 6 else index == 0
                Text(
                    text = dayName,
                    modifier = Modifier.weight(1f),
                    textAlign = TextAlign.Center,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = if (isSunday) dangerColor else primaryColor
                )
            }
        }

        // Calendar Grid
        AnimatedContent(
            targetState = "$currentYear-$currentMonth",
            transitionSpec = {
                if (transitionDirection > 0) {
                    slideInHorizontally(animationSpec = tween(300)) { width -> width } with
                    slideOutHorizontally(animationSpec = tween(300)) { width -> -width }
                } else {
                    slideInHorizontally(animationSpec = tween(300)) { width -> -width } with
                    slideOutHorizontally(animationSpec = tween(300)) { width -> width }
                }.using(SizeTransform(clip = false))
            }
        ) { _ ->
            Column(modifier = Modifier.fillMaxWidth()) {
                if (isLoading) {
                    val chunks = (0 until 42).chunked(7)
                    chunks.forEach { week ->
                        Row(modifier = Modifier.fillMaxWidth()) {
                            week.forEach { _ ->
                                Box(
                                    modifier = Modifier
                                        .weight(1f)
                                        .height(64.dp)
                                        .padding(vertical = 2.dp)
                                ) {
                                    BizSkeletonText(
                                        modifier = Modifier
                                            .fillMaxSize()
                                            .padding(horizontal = 2.dp)
                                            .clip(RoundedCornerShape(4.dp))
                                    )
                                }
                            }
                        }
                    }
                } else {
                    val chunks = days.chunked(7)
                    chunks.forEach { week ->
                        Row(modifier = Modifier.fillMaxWidth()) {
                            week.forEach { day ->
                                val isSelected = when (config.selectionMode) {
                                    SelectionMode.Single -> isSameDay(selectedDate, day.date)
                                    else -> isSameDay(rangeStartDate, day.date) || isSameDay(rangeEndDate, day.date)
                                }
                                val isRangeStart = config.selectionMode != SelectionMode.Single && isSameDay(rangeStartDate, day.date)
                                val isRangeEnd = config.selectionMode != SelectionMode.Single && isSameDay(rangeEndDate, day.date)
                                val isInRange = config.selectionMode != SelectionMode.Single && rangeStartDate != null && rangeEndDate != null &&
                                        day.date.time > rangeStartDate!!.time && day.date.time < rangeEndDate!!.time
    
                                Box(
                                    modifier = Modifier
                                        .weight(1f)
                                        .height(64.dp)
                                        .padding(vertical = 2.dp)
                                        .clickable(enabled = day.isEnabled && day.isCurrentMonth) { onDayClick(day) },
                                    contentAlignment = Alignment.Center
                                ) {
                                    if (!day.isHidden) {
                                        // Range backgrounds
                                        if (isInRange) {
                                            Box(modifier = Modifier.fillMaxSize().background(rangeBgColor))
                                        }
                                        if (isRangeStart && rangeEndDate != null && !isSameDay(rangeStartDate, rangeEndDate)) {
                                            Box(modifier = Modifier.fillMaxSize().padding(start = 32.dp).background(rangeBgColor))
                                        }
                                        if (isRangeEnd && rangeStartDate != null && !isSameDay(rangeStartDate, rangeEndDate)) {
                                            Box(modifier = Modifier.fillMaxSize().padding(end = 32.dp).background(rangeBgColor))
                                        }
    
                                        // Day content background
                                        val contentModifier = Modifier
                                            .fillMaxSize()
                                            .padding(horizontal = 2.dp)
                                            .clip(RoundedCornerShape(4.dp))
                                            .then(
                                                if (isSelected) Modifier.background(dangerColor)
                                                else if (day.isToday && !isSelected) Modifier.border(1.dp, dangerColor, RoundedCornerShape(4.dp))
                                                else Modifier
                                            )
                                            
                                        Box(modifier = contentModifier, contentAlignment = Alignment.Center) {
                                            if (day.isSpecial) {
                                                Box(modifier = Modifier.align(Alignment.TopEnd).padding(4.dp).size(6.dp).background(Color(0xFFFFA000), CircleShape))
                                            }
                                            
                                            Text(
                                                text = day.gregorianDay.toString(),
                                                fontSize = 18.sp,
                                                color = if (isSelected) Color.White 
                                                        else if (!day.isEnabled || !day.isCurrentMonth) primaryColor.copy(alpha = 0.3f)
                                                        else if (day.isSunday) dangerColor 
                                                        else primaryColor
                                            )
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        if (showActionButtons) {
            Row(
                modifier = Modifier.fillMaxWidth().padding(top = 16.dp),
                horizontalArrangement = Arrangement.End
            ) {
                TextButton(onClick = { onCancel?.invoke() }) {
                    Text("Cancel", color = primaryColor, fontWeight = FontWeight.SemiBold)
                }
                Spacer(modifier = Modifier.width(16.dp))
                Button(
                    onClick = { onConfirm?.invoke(if (config.selectionMode == SelectionMode.Single) selectedDate else rangeStartDate, if (config.selectionMode == SelectionMode.Single) null else rangeEndDate) },
                    colors = ButtonDefaults.buttonColors(containerColor = dangerColor)
                ) {
                    Text("Confirm", color = Color.White, fontWeight = FontWeight.SemiBold)
                }
            }
        }
    }
}

File: ios/ui-framework-sample/UIFramework/Sources/UIFramework/BizCalendar.swift
import SwiftUI

public struct CalendarSelection {
    public let date: Date
}

public struct BizCalendar: View {
    public var config: CalendarConfig
    @State public var initialDate: Date
    public var showActionButtons: Bool
    
    public var onDateSelected: ((CalendarSelection) -> Void)?
    public var onRangeSelected: ((CalendarSelection?, CalendarSelection?) -> Void)?
    public var onMonthChanged: ((Int, Int) -> Void)?
    public var onConfirm: ((CalendarSelection?, CalendarSelection?) -> Void)?
    public var onCancel: (() -> Void)?
    
    @State private var selectedDate: Date?
    @State private var rangeStartDate: Date?
    @State private var rangeEndDate: Date?
    
    @State private var currentMonth: Int
    @State private var currentYear: Int
    @State private var showMonthPicker = false
    
    @State private var dragOffset: CGFloat = 0
    @State private var isLoading: Bool = false
    
    @State private var days: [CalendarDay] = []
    
    private let calendarState: CalendarState
    private let monthNames = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
    
    public init(
        config: CalendarConfig = CalendarConfig(),
        initialDate: Date = Date(),
        showActionButtons: Bool = false,
        onDateSelected: ((CalendarSelection) -> Void)? = nil,
        onRangeSelected: ((CalendarSelection?, CalendarSelection?) -> Void)? = nil,
        onMonthChanged: ((Int, Int) -> Void)? = nil,
        onConfirm: ((CalendarSelection?, CalendarSelection?) -> Void)? = nil,
        onCancel: (() -> Void)? = nil
    ) {
        self.config = config
        self._initialDate = State(initialValue: initialDate)
        self.showActionButtons = showActionButtons
        
        self.onDateSelected = onDateSelected
        self.onRangeSelected = onRangeSelected
        self.onMonthChanged = onMonthChanged
        self.onConfirm = onConfirm
        self.onCancel = onCancel
        
        let cal = Calendar.current
        self._currentMonth = State(initialValue: cal.component(.month, from: initialDate))
        self._currentYear = State(initialValue: cal.component(.year, from: initialDate))
        
        self.calendarState = CalendarState(config: config)
        
        self._selectedDate = State(initialValue: initialDate)
    }
    
    private var daysOfWeek: [String] {
        let firstDay = config.firstDayOfWeek
        if firstDay == 1 {
            return ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
        }
        return ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"]
    }
    
    private func isSundayIndex(_ idx: Int) -> Bool {
        return config.firstDayOfWeek == 1 ? idx == 6 : idx == 0
    }
    
    public var body: some View {
        VStack(spacing: 0) {
            headerView
            
            weekdaysView
            
            calendarGridView
            
            if showActionButtons {
                actionsView
            }
        }
        .padding(16)
        .background(Color.white)
        .onAppear {
            updateCalendar()
        }
        .onChange(of: currentMonth) { _ in updateCalendar() }
        .onChange(of: currentYear) { _ in updateCalendar() }
        .onChange(of: config) { _ in updateCalendar() }
        .sheet(isPresented: $showMonthPicker) {
            MonthYearPicker(
                month: $currentMonth,
                year: $currentYear,
                isPresented: $showMonthPicker,
                monthNames: monthNames
            )
            .withPresentationDetents()
        }
    }
    
    private var headerView: some View {
        HStack {
            Button(action: { showMonthPicker = true }) {
                HStack(spacing: 4) {
                    Text("\(monthNames[currentMonth - 1]) \(String(currentYear))")
                        .font(.system(size: 18, weight: .semibold))
                        .foregroundColor(Color(red: 26/255, green: 42/255, blue: 94/255))
                    Image(systemName: "chevron.down")
                        .font(.system(size: 14))
                }
            }
            Spacer()
            HStack(spacing: 8) {
                Button(action: { animateToPrevMonth() }) {
                    Image(systemName: "chevron.left")
                        .frame(width: 24, height: 24)
                }
                Button(action: goToToday) {
                    Image(systemName: "calendar")
                        .frame(width: 24, height: 24)
                        .foregroundColor(Color(red: 0, green: 51/255, blue: 153/255))
                }
                Button(action: { animateToNextMonth() }) {
                    Image(systemName: "chevron.right")
                        .frame(width: 24, height: 24)
                }
            }
            .foregroundColor(.primary)
        }
        .padding(.vertical, 16)
    }
    
    private var weekdaysView: some View {
        HStack(spacing: 0) {
            ForEach(0..<7, id: \.self) { idx in
                Text(daysOfWeek[idx])
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundColor(isSundayIndex(idx) ? Color.red : Color(red: 26/255, green: 42/255, blue: 94/255))
                    .frame(maxWidth: .infinity)
            }
        }
        .padding(.bottom, 8)
    }
    
    private var calendarGridView: some View {
        let columns = Array(repeating: GridItem(.flexible(), spacing: 0), count: 7)
        return LazyVGrid(columns: columns, spacing: 0) {
            if isLoading {
                ForEach(0..<42, id: \.self) { _ in
                    skeletonCell
                }
            } else {
                ForEach(days) { day in
                    dayCell(for: day)
                }
            }
        }
        .offset(x: dragOffset)
        .gesture(
            DragGesture()
                .onChanged { value in
                    dragOffset = value.translation.width
                }
                .onEnded { value in
                    if value.translation.width > 50 {
                        animateToPrevMonth()
                    } else if value.translation.width < -50 {
                        animateToNextMonth()
                    } else {
                        withAnimation { dragOffset = 0 }
                    }
                }
        )
    }
    
    private var skeletonCell: some View {
        ZStack {
            BizSkeletonText()
                .padding(2)
        }
        .frame(height: 64)
    }
    
    private func dayCell(for day: CalendarDay) -> some View {
        ZStack {
            if day.isHidden {
                Color.clear
            } else {
                let isDisabled = !day.isEnabled || !day.isCurrentMonth
                let selected = isSelected(day)
                let inRange = isInRange(day)
                let rangeStart = isRangeStart(day)
                let rangeEnd = isRangeEnd(day)
                
                ZStack {
                    if inRange || rangeStart || rangeEnd {
                        HStack(spacing: 0) {
                            if rangeStart && !rangeEnd {
                                Color.clear.frame(maxWidth: .infinity)
                                Color.red.opacity(0.1).frame(maxWidth: .infinity)
                            } else if rangeEnd && !rangeStart {
                                Color.red.opacity(0.1).frame(maxWidth: .infinity)
                                Color.clear.frame(maxWidth: .infinity)
                            } else if inRange {
                                Color.red.opacity(0.1)
                            } else {
                                Color.clear
                            }
                        }
                    }
                    
                    VStack {
                        if day.isSpecial {
                            Circle()
                                .fill(Color.orange)
                                .frame(width: 6, height: 6)
                                .position(x: 32, y: 8) // Approx position
                        }
                        Text("\(day.gregorianDay)")
                            .font(.system(size: 18))
                            .foregroundColor(
                                selected ? .white :
                                    (day.isSunday ? .red :
                                        (isDisabled ? Color.gray.opacity(0.5) : Color.primary))
                            )
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                    .background(
                        selected ? Color.red : Color.clear
                    )
                    .cornerRadius(4)
                    .overlay(
                        RoundedRectangle(cornerRadius: 4)
                            .stroke(day.isToday && !selected ? Color.red : Color.clear, lineWidth: 1)
                    )
                    .padding(2)
                }
                .frame(height: 64)
                .opacity(isDisabled ? 0.3 : 1.0)
                .onTapGesture {
                    onDayClick(day)
                }
            }
        }
        .frame(height: 64)
    }
    
    private var actionsView: some View {
        HStack(spacing: 16) {
            Spacer()
            Button("Cancel") {
                onCancel?()
            }
            .font(.system(size: 14, weight: .semibold))
            .foregroundColor(Color(red: 26/255, green: 42/255, blue: 94/255))
            
            Button("Confirm") {
                onConfirmAction()
            }
            .font(.system(size: 14, weight: .semibold))
            .foregroundColor(.white)
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
            .background(Color.red)
            .cornerRadius(4)
        }
        .padding(.top, 16)
    }
    
    private func updateCalendar() {
        isLoading = true
        let year = currentYear
        let month = currentMonth
        
        DispatchQueue.global(qos: .userInitiated).async {
            let newDays = self.calendarState.generateDaysForMonth(year: year, month: month)
            
            DispatchQueue.main.async {
                self.days = newDays
                self.isLoading = false
                self.onMonthChanged?(year, month)
            }
        }
    }
    
    private func animateToPrevMonth() {
        withAnimation(.easeInOut(duration: 0.2)) {
            dragOffset = UIScreen.main.bounds.width
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
            prevMonth()
            dragOffset = -UIScreen.main.bounds.width
            withAnimation(.easeInOut(duration: 0.2)) {
                dragOffset = 0
            }
        }
    }
    
    private func animateToNextMonth() {
        withAnimation(.easeInOut(duration: 0.2)) {
            dragOffset = -UIScreen.main.bounds.width
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
            nextMonth()
            dragOffset = UIScreen.main.bounds.width
            withAnimation(.easeInOut(duration: 0.2)) {
                dragOffset = 0
            }
        }
    }
    
    private func prevMonth() {
        if currentMonth == 1 {
            currentMonth = 12
            currentYear -= 1
        } else {
            currentMonth -= 1
        }
    }
    
    private func nextMonth() {
        if currentMonth == 12 {
            currentMonth = 1
            currentYear += 1
        } else {
            currentMonth += 1
        }
    }
    
    private func goToToday() {
        let cal = Calendar.current
        let today = Date()
        currentMonth = cal.component(.month, from: today)
        currentYear = cal.component(.year, from: today)
    }
    
    private func isSameDay(_ d1: Date?, _ d2: Date?) -> Bool {
        guard let d1 = d1, let d2 = d2 else { return false }
        let cal = Calendar.current
        return cal.isDate(d1, inSameDayAs: d2)
    }
    
    private func isSelected(_ day: CalendarDay) -> Bool {
        if config.selectionMode == .single {
            return isSameDay(selectedDate, day.date)
        }
        return isSameDay(rangeStartDate, day.date) || isSameDay(rangeEndDate, day.date)
    }
    
    private func isInRange(_ day: CalendarDay) -> Bool {
        guard config.selectionMode != .single, let start = rangeStartDate, let end = rangeEndDate else { return false }
        return day.date > start && day.date < end
    }
    
    private func isRangeStart(_ day: CalendarDay) -> Bool {
        if config.selectionMode == .single { return false }
        return isSameDay(rangeStartDate, day.date)
    }
    
    private func isRangeEnd(_ day: CalendarDay) -> Bool {
        if config.selectionMode == .single { return false }
        return isSameDay(rangeEndDate, day.date)
    }
    
    private func onDayClick(_ day: CalendarDay) {
        if !day.isEnabled || !day.isCurrentMonth { return }
        
        if config.selectionMode == .single {
            selectedDate = day.date
            onDateSelected?(CalendarSelection(date: day.date))
        } else if config.selectionMode == .week {
            if let index = days.firstIndex(where: { isSameDay($0.date, day.date) }) {
                let rowStart = (index / 7) * 7
                let weekDays = days[rowStart..<min(rowStart + 7, days.count)].filter { $0.isEnabled && !$0.isHidden }
                if let first = weekDays.first, let last = weekDays.last {
                    rangeStartDate = first.date
                    rangeEndDate = last.date
                    onRangeSelected?(CalendarSelection(date: first.date), CalendarSelection(date: last.date))
                }
            }
        } else {
            if rangeStartDate == nil || (rangeStartDate != nil && rangeEndDate != nil) {
                rangeStartDate = day.date
                rangeEndDate = nil
            } else if let start = rangeStartDate, rangeEndDate == nil {
                if day.date < start {
                    rangeStartDate = day.date
                } else {
                    rangeEndDate = day.date
                    onRangeSelected?(CalendarSelection(date: start), CalendarSelection(date: day.date))
                }
            }
        }
    }
    
    private func onConfirmAction() {
        let startDay = days.first { isSameDay($0.date, config.selectionMode == .single ? selectedDate : rangeStartDate) }
        let endDay = days.first { isSameDay($0.date, rangeEndDate) }
        
        if config.selectionMode == .single {
            onConfirm?(startDay.map { CalendarSelection(date: $0.date) }, nil)
        } else {
            onConfirm?(
                startDay.map { CalendarSelection(date: $0.date) },
                endDay.map { CalendarSelection(date: $0.date) }
            )
        }
    }
}

struct MonthYearPicker: View {
    @Binding var month: Int
    @Binding var year: Int
    @Binding var isPresented: Bool
    var monthNames: [String]
    
    @State private var tempMonth: Int
    @State private var tempYear: Int
    
    private let yearRange: [Int]
    
    init(month: Binding<Int>, year: Binding<Int>, isPresented: Binding<Bool>, monthNames: [String]) {
        self._month = month
        self._year = year
        self._isPresented = isPresented
        self.monthNames = monthNames
        
        let currentYear = Calendar.current.component(.year, from: Date())
        self.yearRange = Array((currentYear - 50)...(currentYear + 50))
        
        self._tempMonth = State(initialValue: month.wrappedValue)
        self._tempYear = State(initialValue: year.wrappedValue)
    }
    
    var body: some View {
        VStack {
            Text("Select Month")
                .font(.headline)
                .padding()
            
            HStack {
                Picker("Month", selection: $tempMonth) {
                    ForEach(1...12, id: \.self) { m in
                        Text(monthNames[m - 1]).tag(m)
                    }
                }
                .pickerStyle(WheelPickerStyle())
                
                Picker("Year", selection: $tempYear) {
                    ForEach(yearRange, id: \.self) { y in
                        Text(String(y)).tag(y)
                    }
                }
                .pickerStyle(WheelPickerStyle())
            }
            .frame(height: 200)
            
            HStack(spacing: 16) {
                Button("Close") {
                    isPresented = false
                }
                .frame(maxWidth: .infinity)
                .padding()
                .background(Color(white: 0.95))
                .cornerRadius(8)
                .foregroundColor(.black)
                
                Button("Confirm") {
                    month = tempMonth
                    year = tempYear
                    isPresented = false
                }
                .frame(maxWidth: .infinity)
                .padding()
                .background(Color(red: 0, green: 51/255, blue: 153/255))
                .cornerRadius(8)
                .foregroundColor(.white)
            }
            .padding()
        }
    }
}

struct PresentationDetentModifier: ViewModifier {
    func body(content: Content) -> some View {
        if #available(iOS 16.0, *) {
            content.presentationDetents([.medium, .fraction(0.4)])
        } else {
            content
        }
    }
}

extension View {
    func withPresentationDetents() -> some View {
        self.modifier(PresentationDetentModifier())
    }
}
